octopusdeploy_framework/datasource_tenant_projects.go

```diff
@@ -0,0 +1,201 @@
+package octopusdeploy_framework
+
+import (
+	"context"
+	"fmt"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/client"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/tenants"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/schemas"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	datasourceSchema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"golang.org/x/exp/slices"
+)
+
+type TenantProjectsDataModel struct {
+	SpaceID        types.String `tfsdk:"space_id"`
+	TenantIDs      types.List   `tfsdk:"tenant_ids"`
+	ProjectIDs     types.List   `tfsdk:"project_ids"`
+	EnvironmentIDs types.List   `tfsdk:"environment_ids"`
+	TenantProjects types.List   `tfsdk:"tenant_projects"`
+}
+
+type tenantProjectsDataSource struct {
+	*Config
+}
+
+func NewTenantProjectDataSource() datasource.DataSource {
+	return &tenantProjectsDataSource{}
+}
+
+func (*tenantProjectsDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = util.GetTypeName("tenant_projects")
+}
+
+func (t *tenantProjectsDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	t.Config = DataSourceConfiguration(req, resp)
+}
+
+func (*tenantProjectsDataSource) Schema(_ context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = datasourceSchema.Schema{
+		Description: "Provides information about existing tenants.",
+		Attributes: map[string]datasourceSchema.Attribute{
+			"tenant_ids":      schemas.GetQueryIDsDatasourceSchema(),
+			"project_ids":     schemas.GetQueryIDsDatasourceSchema(),
+			"environment_ids": schemas.GetQueryIDsDatasourceSchema(),
+			"space_id":        schemas.GetSpaceIdDatasourceSchema("tenant projects", false),
```

Comment:
Our other DataSources provide paged results sets. 

Is it worth trying to do so / would it be possible with the shape of the API?

Reply:
paging is hard as I filter local and the data structure we are working with is a "map". Go does not guarantee the order of the items in a map. In the end, its to hard with our API, will still need to fetch everything to filter it and make the "pages" anyway.